### PR TITLE
Feature add DeiT val and training

### PR DESCRIPTION
**Pull request by xperzy**
Impl deit model
Impl regnety_160 model
add validation 
add training
add auto_augment, mixup, cutmix, label smoothing
add distillation training

### CODE REVIEW

File: image_classification/DeiT/datasets.py [hunk 1]
```diff
@@ -7,9 +21,12 @@
 import math
 from paddle.io import Dataset, DataLoader, DistributedBatchSampler
 from paddle.vision import transforms, datasets, image_load
+from auto_augment import auto_augment_policy_original
+from auto_augment import AutoAugment
+from random_erasing import RandomErasing
 
 class ImageNet2012Dataset(Dataset):
-    """Build ImageNet2012 dataset
+    """Build ImageNet2021 dataset
```
Comment (jonny4929) [inline image_classification/DeiT/datasets.py:29]: 2012 or 2021?

File: image_classification/DeiT/datasets.py [hunk 2]
```diff
@@ -21,7 +38,7 @@ class ImageNet2012Dataset(Dataset):
     """
 
     def __init__(self, file_folder, mode="train", transform=None):
-        """Init ImageNet2012 Dataset with dataset file path, mode(train/val), and transform"""
+        """Init ImageNet2021Dataset with dataset file path, mode(train/val), and transform"""
```
Comment (jonny4929) [inline image_classification/DeiT/datasets.py:41]: 2012 or 2021?

File: image_classification/DeiT/load_pytorch_weights_base_patch16_384.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+import numpy as np
+import paddle
+import torch
+from deit import *
+from config import *
+
+config = get_config()
+parser = argparse.ArgumentParser('')
+parser.add_argument('-cfg', type=str, default='./configs/deit_base_patch16_384.yaml')
+parser.add_argument('-dataset', type=str, default=None)
```
Comment (jonny4929) [inline image_classification/DeiT/load_pytorch_weights_base_patch16_384.py:25]: the following args are useless

File: image_classification/DeiT/load_pytorch_weights_base_patch16_224.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import numpy as np
+import paddle
+import torch
+from deit import *
+from config import *
+
+config = get_config()
+parser = argparse.ArgumentParser('')
+parser.add_argument('-cfg', type=str, default='./configs/deit_base_patch16_224.yaml')
+parser.add_argument('-dataset', type=str, default=None)
```
Comment (jonny4929) [inline image_classification/DeiT/load_pytorch_weights_base_patch16_224.py:25]: the followingargs are useless

File: image_classification/DeiT/load_pytorch_weights_base_patch16_224.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    out_torch = out_torch.data.cpu().numpy()
+    out_paddle = out_paddle.cpu().numpy()
+
+    print(out_torch.shape, out_paddle.shape)
+    print(out_torch[0, 0:100])
+    print(out_paddle[0, 0:100])
+    assert np.allclose(out_torch, out_paddle, atol = 1e-5)
+    
+    # save weights for paddle model
+    model_path = os.path.join('./deit_base_distilled_patch16_224.pdparams')
```
Comment (jonny4929) [inline image_classification/DeiT/load_pytorch_weights_base_patch16_224.py:170]: why using os.path.join

Comment (jonny4929) [reply image_classification/DeiT/load_pytorch_weights_base_patch16_224.py:170]: also for image_classification/DeiT/load_pytorch_weights_regnety_160.py and image_classification/DeiT/load_pytorch_weights_base_patch16_384.py

File: image_classification/DeiT/model_ema.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    def __init__(self, model, decay=0.999):
+        self.module = copy.deepcopy(model)
+        self.module.eval()
+        self.decay = decay
+
+    @paddle.no_grad()
+    def _update(self, model, update_fn):
+        for ema_param_values, model_param_values in zip(
+            self.module.state_dict().values(), model.state_dict().values()):
+            ema_param_values = copy.deepcopy(update_fn(ema_param_values, model_param_values))
```
Comment (jonny4929) [inline image_classification/DeiT/model_ema.py:45]: this operation doesn't make sense!

File: image_classification/DeiT/tests/test_ema.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            target = paddle.randint(0, 5, [4])
+            out = model(x)
+            loss = criterion(out, target)
+            loss.backward()
+            optim.step()
+            optim.clear_grad()
+            model_ema.update(model)
+            # test model ema update
+            model_ema_weight = model_ema.module.head.weight
+            self.assertFalse(np.allclose(prev_weight.numpy(), model_ema_weight.numpy(), atol=1e-5))
```
Comment (jonny4929) [inline image_classification/DeiT/tests/test_ema.py:62]: prev_weight is not assigned in the first loop

File: image_classification/DeiT/mixup.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        self.switch_prob = switch_prob
+        self.label_smoothing = label_smoothing
+        self.num_classes = num_classes
+        self.mode = mode
+        self.correct_lam = correct_lam
+        assert mode == 'batch', 'Now only batch mode is supported!'
+
+    def __call__(self, x, target):
+        assert x.shape[0] % 2 == 0, "Batch size should be even"
+        lam = self._mix_batch(x)
```
Comment (jonny4929) [inline image_classification/DeiT/mixup.py:174]: cutmix is useable, but mixup without cutmix makes no difference

Comment (jonny4929) [reply image_classification/DeiT/mixup.py:174]: the assignment in line 215 may be the reason

File: image_classification/DeiT/main_single_gpu.py [hunk 9]
```diff
@@ -102,8 +133,15 @@ def train(dataloader,
...
 
-        pred = F.softmax(output)
-        acc = paddle.metric.accuracy(pred, label.unsqueeze(1))
+        if model_ema is not None:
+            model_ema.update(model)
+
+        # average of output and kd_output, like model eval mode
+        pred = F.softmax((output[0] + output[1]) / 2)
+        if mixup_fn:
+            acc = paddle.metric.accuracy(pred, label_orig)
```
Comment (jonny4929) [inline image_classification/DeiT/main_single_gpu.py:142]: the api only accept int64 data in python3.7, paddle2.1.0, though the doc shows it accepts both int64 and int32
